### PR TITLE
Coding guidelines update

### PR DESCRIPTION
**Pull request by maartenba**
* Re-wrote coding guidelines based on aspnet
* Added R# settings file which we can use on all projects

@NuGet/core-team, please review

### CODE REVIEW

File: NuGet.Docs/Contribute/Coding-Guidelines.md [hunk 1]
```diff
@@ -1,102 +1,511 @@
...
+
+### Repos
+
+To create a new repo in the https://github.com/nuget/ org, contact @yishaigalatzer.
+
+### Home repo
+
+The https://github.com/nuget/Home repo is the starting point for all things NuGet. It has the [issue tracker](https://github.com/nuget/home/issues) as well as an [up-to-date README](https://github.com/NuGet/Home/blob/master/README.md) of how repos are structured. Make sure to consult it before beginning your journey through NuGet code.
+
+### Branch strategy
```
Comment (maartenba) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: Are we going to adopt these?

Comment (davidfowl) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: It's the only sane thing to do :smile:

File: NuGet.Docs/Contribute/Coding-Guidelines.md [hunk 2]
```diff
@@ -1,102 +1,511 @@
...
+
+Shortly before a release, the `release` branches are created from `dev`, and stabilization work happens there. Post-release work continus in the `dev` branch. Once the release takes place, the code is pushed from `release` to `master`.
+
+### Solution and project folder structure and naming
+
+Solution files generally go in the repo root.
+
+Solution names match repo names (e.g. NuGet.sln in the NuGet repo).
+
+Every project also needs a `project.json` and a matching `.kproj` file. This `project.json` is the source of truth for a project's dependencies and configuration options.
```
Comment (anurse) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: `kproj` => `xproj`

File: NuGet.Docs/Contribute/Coding-Guidelines.md [hunk 3]
```diff
@@ -1,102 +1,511 @@
...
+All these files are set to be ignored in the `.gitignore` file.
+
+### Conditional compilation for Desktop/CoreCLR
+
+Almost all development is done for both CoreCLR and Desktop .NET. Some code will be CoreCLR-specific or Desktop-specific because of API changes or behavior differences. The build system has two conditional compilation statements to assist with this:
+
+Desktop:
+
+```c#
+#ifdef ASPNET50
```
Comment (anurse) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: `ASPNET50` => `DNX451` or `NET451` (depending on exact target)

File: NuGet.Docs/Contribute/Coding-Guidelines.md [hunk 4]
```diff
@@ -1,102 +1,511 @@
...
+```
+
+CoreCLR:
+
+```c#
+#ifdef ASPNETCORE50
```
Comment (anurse) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: `ASPNETCORE50` => `DNXCORE50`

**Comment by anurse** [comment]
Minor things, though if you copied them from `aspnet`, it may be that we need to update the docs there too :).

Should we also call out explicit guidance on optional parameters on public APIs? I've heard drastically conflicting policies from various people.

Looks good!

File: NuGet.Docs/Contribute/Coding-Guidelines.md (hunk 2 above)
Comment (maartenba) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: Fixed in https://github.com/NuGet/NuGetDocs/commit/81004c2142b59f7eb84617789fdffe153d330bb5. Needs an update in https://github.com/aspnet/Home/wiki/Engineering-guidelines as well.

File: NuGet.Docs/Contribute/Coding-Guidelines.md (hunk 3 above)
Comment (maartenba) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: Fixed in https://github.com/NuGet/NuGetDocs/commit/81004c2142b59f7eb84617789fdffe153d330bb5. Needs an update in https://github.com/aspnet/Home/wiki/Engineering-guidelines as well.

File: NuGet.Docs/Contribute/Coding-Guidelines.md (hunk 4 above)
Comment (maartenba) [inline NuGet.Docs/Contribute/Coding-Guidelines.md]: Fixed in https://github.com/NuGet/NuGetDocs/commit/81004c2142b59f7eb84617789fdffe153d330bb5. Needs an update in https://github.com/aspnet/Home/wiki/Engineering-guidelines as well.

**Comment by maartenba** [comment]
Thanks @anurse! We will indeed need some aspnet wiki updates, too.

**Comment by maartenba** [comment]
With no additional comments, merging it in.

**Pull request merged**